paxml/trainer_lib.py

```diff
@@ -1130,17 +1130,41 @@ def train_step_single_learner(
     wps_with_opt = tasks_lib.filter_vars_for_grad_or_opt(
         var_weight_hparams, excluded_for_opt
     )
+
+    # Except for the overwrite_with_gradient params, we use the update_states()
+    # to compute the optimizer states and transformed grads.
+    overwrite_var_mask = \
+        tasks_lib.get_overwrite_with_gradient_var_mask(var_weight_hparams)
+    grads_with_opt = tasks_lib.filter_vars_for_grad_or_opt(
+        grads, overwrite_var_mask
+    )
     transformed_grads, new_opt_states = learner.update_states(
```

Comment:
Let's make the update_states and apply_gradient API more uniform. I.e. either both should be aware of OWG variables (and then the learner is fully responsible for OWG). Or none should, and then trainer is fully responsible for OWG.

Reply:
Done.